## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 427717`
Commit: [`d2b06c65021e65df768ba967750873ce494adb39`](https://github.com/NixOS/nixpkgs/commit/d2b06c65021e65df768ba967750873ce494adb39) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/d2b06c65021e65df768ba967750873ce494adb39..pull/427717/head))
Merge: [`c897d8f4abfc58c155dd9f75ddf6d6e8af97f98c`](https://github.com/NixOS/nixpkgs/commit/c897d8f4abfc58c155dd9f75ddf6d6e8af97f98c)

Logs: https://github.com/Defelo/nixpkgs-review-gha/actions/runs/16468715507

<details><summary>Download packages from cache:</summary><ul>
<li><details><summary><code>x86_64-linux</code></summary>

```shell
nix-store -r --add-root nixpkgs-pr-427717-x86_64-linux \
  --option binary-caches 'https://cache.nixos.org/ https://attic.defelo.de/nixpkgs' \
  --option trusted-public-keys '
  cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=
  nixpkgs:xeaAWa3crK09hMmFiygBeRmLq3hUjUShgaAwYVUEtw0=
  ' \
  /nix/store/3s77vbijls7sbdwzz0xji9ddfxzkwzy2-go-httpbin-2.18.3
```
</details></li>
</ul></details>


---
### `x86_64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>go-httpbin</li>
  </ul>
</details>


---
### `aarch64-linux`
<details>
  <summary>:fast_forward: 2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>go-httpbin</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `aarch64-linux`</summary>
<details>
<summary>go-httpbin</summary>
<pre>@nix { &quot;action&quot;: &quot;setPhase&quot;, &quot;phase&quot;: &quot;updateAutotoolsGnuConfigScriptsPhase&quot; }
Running phase: configurePhase
@nix { &quot;action&quot;: &quot;setPhase&quot;, &quot;phase&quot;: &quot;configurePhase&quot; }
Running phase: buildPhase
@nix { &quot;action&quot;: &quot;setPhase&quot;, &quot;phase&quot;: &quot;buildPhase&quot; }
Building subPackage ./cmd/go-httpbin
Building subPackage ./httpbin
Building subPackage ./httpbin/cmd
Building subPackage ./httpbin/digest
Building subPackage ./httpbin/websocket
Building subPackage ./internal/testing/assert
Building subPackage ./internal/testing/must
Running phase: checkPhase
@nix { &quot;action&quot;: &quot;setPhase&quot;, &quot;phase&quot;: &quot;checkPhase&quot; }
--- FAIL: TestSSE (0.00s)
    --- FAIL: TestSSE/handle_cancelation_during_stream (0.10s)
        handlers_test.go:3527: error making HTTP request: GET http://127.0.0.1:34161/sse?duration=900ms&amp;delay=0&amp;count=2: Get &quot;http://127.0.0.1:34161/sse?duration=900ms&amp;delay=0&amp;count=2&quot;: context deadline exceeded
FAIL
FAIL	github.com/mccutchen/go-httpbin/v2/httpbin	4.513s
FAIL</pre>
</details>
</details>


---
### `x86_64-darwin` (sandbox = true)
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>go-httpbin</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `x86_64-darwin`</summary>
<details>
<summary>go-httpbin</summary>
<pre>@nix { &quot;action&quot;: &quot;setPhase&quot;, &quot;phase&quot;: &quot;checkPhase&quot; }
panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: bind: operation not permitted

goroutine 1 [running]:
net/http/httptest.newLocalListener()
	/nix/store/b56ngr9p7x6z368w8m7ps95r3x8gdfv7-go-1.24.4/share/go/src/net/http/httptest/server.go:71 +0xda
net/http/httptest.NewUnstartedServer(...)
	/nix/store/b56ngr9p7x6z368w8m7ps95r3x8gdfv7-go-1.24.4/share/go/src/net/http/httptest/server.go:119
net/http/httptest.NewServer({0x63adac0, 0xc0000dc000})
	/nix/store/b56ngr9p7x6z368w8m7ps95r3x8gdfv7-go-1.24.4/share/go/src/net/http/httptest/server.go:106 +0x25
github.com/mccutchen/go-httpbin/v2/httpbin.newTestServer({0x63adac0?, 0xc0000dc000?})
	/nix/build/nix-build-go-httpbin-2.18.3.drv-0/source/httpbin/handlers_test.go:3676 +0x1d
github.com/mccutchen/go-httpbin/v2/httpbin.newTestEnvironment(...)
	/nix/build/nix-build-go-httpbin-2.18.3.drv-0/source/httpbin/handlers_test.go:3687
github.com/mccutchen/go-httpbin/v2/httpbin.TestMain(0xc00008a8c0)
	/nix/build/nix-build-go-httpbin-2.18.3.drv-0/source/httpbin/handlers_test.go:85 +0x90
main.main()
	_testmain.go:183 +0xa8
FAIL	github.com/mccutchen/go-httpbin/v2/httpbin	0.567s
FAIL</pre>
</details>
</details>


---
### `aarch64-darwin` (sandbox = true)
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>go-httpbin</li>
  </ul>
</details>

---
<details>
<summary>Error logs: `aarch64-darwin`</summary>
<details>
<summary>go-httpbin</summary>
<pre>@nix { &quot;action&quot;: &quot;setPhase&quot;, &quot;phase&quot;: &quot;checkPhase&quot; }
panic: httptest: failed to listen on a port: listen tcp6 [::1]:0: bind: operation not permitted

goroutine 1 [running]:
net/http/httptest.newLocalListener()
	/nix/store/f4a0g1p943l61wfvqnpdr73v9bsyfhf2-go-1.24.4/share/go/src/net/http/httptest/server.go:71 +0xdc
net/http/httptest.NewUnstartedServer(...)
	/nix/store/f4a0g1p943l61wfvqnpdr73v9bsyfhf2-go-1.24.4/share/go/src/net/http/httptest/server.go:119
net/http/httptest.NewServer({0x101227740, 0x14000152000})
	/nix/store/f4a0g1p943l61wfvqnpdr73v9bsyfhf2-go-1.24.4/share/go/src/net/http/httptest/server.go:106 +0x28
github.com/mccutchen/go-httpbin/v2/httpbin.newTestServer({0x101227740?, 0x14000152000?})
	/nix/build/nix-build-go-httpbin-2.18.3.drv-0/source/httpbin/handlers_test.go:3676 +0x24
github.com/mccutchen/go-httpbin/v2/httpbin.newTestEnvironment(...)
	/nix/build/nix-build-go-httpbin-2.18.3.drv-0/source/httpbin/handlers_test.go:3687
github.com/mccutchen/go-httpbin/v2/httpbin.TestMain(0x14000108780)
	/nix/build/nix-build-go-httpbin-2.18.3.drv-0/source/httpbin/handlers_test.go:85 +0x98
main.main()
	_testmain.go:183 +0x98
FAIL	github.com/mccutchen/go-httpbin/v2/httpbin	0.025s
FAIL</pre>
</details>
</details>

--- a/nixos/doc/manual/release-notes/rl-2511.section.md
+++ b/nixos/doc/manual/release-notes/rl-2511.section.md
@@ -38,6 +38,8 @@
 
 - [Chhoto URL](https://github.com/SinTan1729/chhoto-url), a simple, blazingly fast, selfhosted URL shortener with no unnecessary features, written in Rust. Available as [services.chhoto-url](#opt-services.chhoto-url.enable).
 
+- [go-httpbin](https://github.com/mccutchen/go-httpbin), a reasonably complete and well-tested golang port of httpbin, with zero dependencies outside the go stdlib. Available as [services.go-httpbin](#opt-services.go-httpbin.enable).
+
 - [tuwunel](https://matrix-construct.github.io/tuwunel/), a federated chat server implementing the Matrix protocol, forked from Conduwuit. Available as [services.matrix-tuwunel](#opt-services.matrix-tuwunel.enable).
 
 - [Broadcast Box](https://github.com/Glimesh/broadcast-box), a WebRTC broadcast server. Available as [services.broadcast-box](options.html#opt-services.broadcast-box.enable).

--- a/nixos/modules/module-list.nix
+++ b/nixos/modules/module-list.nix
@@ -1575,6 +1575,7 @@
   ./services/web-apps/gerrit.nix
   ./services/web-apps/glance.nix
   ./services/web-apps/glitchtip.nix
+  ./services/web-apps/go-httpbin.nix
   ./services/web-apps/goatcounter.nix
   ./services/web-apps/gotify-server.nix
   ./services/web-apps/gotosocial.nix

--- a/nixos/modules/services/web-apps/go-httpbin.nix
+++ b/nixos/modules/services/web-apps/go-httpbin.nix
@@ -1,0 +1,111 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.services.go-httpbin;
+
+  environment = lib.mapAttrs (
+    _: value: if lib.isBool value then lib.boolToString value else toString value
+  ) cfg.settings;
+in
+
+{
+  meta.maintainers = with lib.maintainers; [ defelo ];
+
+  options.services.go-httpbin = {
+    enable = lib.mkEnableOption "go-httpbin";
+
+    package = lib.mkPackageOption pkgs "go-httpbin" { };
+
+    settings = lib.mkOption {
+      description = ''
+        Configuration of go-httpbin.
+        See <https://github.com/mccutchen/go-httpbin#configuration> for a list of options.
+      '';
+      example = {
+        HOST = "0.0.0.0";
+        PORT = 8080;
+      };
+
+      type = lib.types.submodule {
+        freeformType =
+          with lib.types;
+          attrsOf (oneOf [
+            str
+            int
+            bool
+          ]);
+
+        options = {
+          HOST = lib.mkOption {
+            type = lib.types.str;
+            description = "The host to listen on.";
+            default = "127.0.0.1";
+            example = "0.0.0.0";
+          };
+
+          PORT = lib.mkOption {
+            type = lib.types.port;
+            description = "The port to listen on.";
+            example = 8080;
+          };
+        };
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.go-httpbin = {
+      wantedBy = [ "multi-user.target" ];
+
+      inherit environment;
+
+      serviceConfig = {
+        User = "go-httpbin";
+        Group = "go-httpbin";
+        DynamicUser = true;
+
+        ExecStart = lib.getExe cfg.package;
+
+        # hardening
+        AmbientCapabilities = "";
+        CapabilityBoundingSet = [ "" ];
+        DevicePolicy = "closed";
+        LockPersonality = true;
+        MemoryDenyWriteExecute = true;
+        NoNewPrivileges = true;
+        PrivateDevices = true;
+        PrivateTmp = true;
+        PrivateUsers = true;
+        ProcSubset = "pid";
+        ProtectClock = true;
+        ProtectControlGroups = true;
+        ProtectHome = true;
+        ProtectHostname = true;
+        ProtectKernelLogs = true;
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        ProtectProc = "invisible";
+        ProtectSystem = "strict";
+        RemoveIPC = true;
+        RestrictAddressFamilies = [ "AF_INET AF_INET6" ];
+        RestrictNamespaces = true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        SocketBindAllow = "tcp:${toString cfg.settings.PORT}";
+        SocketBindDeny = "any";
+        SystemCallArchitectures = "native";
+        SystemCallFilter = [
+          "@system-service"
+          "~@privileged"
+          "~@resources"
+        ];
+        UMask = "0077";
+      };
+    };
+  };
+}

--- a/nixos/tests/all-tests.nix
+++ b/nixos/tests/all-tests.nix
@@ -607,6 +607,7 @@ in
   gnupg = runTest ./gnupg.nix;
   goatcounter = runTest ./goatcounter.nix;
   go-camo = runTest ./go-camo.nix;
+  go-httpbin = runTest ./go-httpbin.nix;
   go-neb = runTest ./go-neb.nix;
   gobgpd = runTest ./gobgpd.nix;
   gocd-agent = runTest ./gocd-agent.nix;

--- a/nixos/tests/go-httpbin.nix
+++ b/nixos/tests/go-httpbin.nix
@@ -1,0 +1,38 @@
+{ lib, ... }:
+
+{
+  name = "go-httpbin";
+  meta.maintainers = with lib.maintainers; [ defelo ];
+
+  nodes.machine = {
+    services.go-httpbin = {
+      enable = true;
+      settings.PORT = 8000;
+    };
+  };
+
+  interactive.nodes.machine = {
+    services.go-httpbin.settings.HOST = "0.0.0.0";
+    networking.firewall.allowedTCPPorts = [ 8000 ];
+    virtualisation.forwardPorts = [
+      {
+        from = "host";
+        host.port = 8000;
+        guest.port = 8000;
+      }
+    ];
+  };
+
+  testScript = ''
+    import json
+
+    machine.wait_for_unit("go-httpbin.service")
+    machine.wait_for_open_port(8000)
+
+    resp = json.loads(machine.succeed("curl localhost:8000/get?foo=bar"))
+    assert resp["args"]["foo"] == ["bar"]
+    assert resp["method"] == "GET"
+    assert resp["origin"] == "127.0.0.1"
+    assert resp["url"] == "http://localhost:8000/get?foo=bar"
+  '';
+}

--- a/pkgs/by-name/go/go-httpbin/package.nix
+++ b/pkgs/by-name/go/go-httpbin/package.nix
@@ -1,0 +1,48 @@
+{
+  lib,
+  buildGoModule,
+  fetchFromGitHub,
+  nix-update-script,
+}:
+
+buildGoModule (finalAttrs: {
+  pname = "go-httpbin";
+  version = "2.18.3";
+
+  src = fetchFromGitHub {
+    owner = "mccutchen";
+    repo = "go-httpbin";
+    tag = "v${finalAttrs.version}";
+    hash = "sha256-ixEbmppQ+4Udc5ytV4YPOpOT/iEbhjQIYGoOGL0dIw8=";
+  };
+
+  vendorHash = null;
+
+  env.CGO_ENABLED = 0;
+
+  ldflags = [
+    "-s"
+    "-w"
+  ];
+
+  # tests are flaky
+  doCheck = false;
+
+  doInstallCheck = true;
+  installCheckPhase = ''
+    runHook preInstallCheck
+    $out/bin/go-httpbin --help &> /dev/null
+    runHook postInstallCheck
+  '';
+
+  passthru.updateScript = nix-update-script { };
+
+  meta = {
+    description = "Reasonably complete and well-tested golang port of httpbin, with zero dependencies outside the go stdlib";
+    homepage = "https://github.com/mccutchen/go-httpbin";
+    changelog = "https://github.com/mccutchen/go-httpbin/releases/tag/v${finalAttrs.version}";
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [ defelo ];
+    mainProgram = "go-httpbin";
+  };
+})

--- a/pkgs/by-name/go/go-httpbin/package.nix
+++ b/pkgs/by-name/go/go-httpbin/package.nix
@@ -2,6 +2,7 @@
   lib,
   buildGoModule,
   fetchFromGitHub,
+  nixosTests,
   nix-update-script,
 }:
 
@@ -35,7 +36,10 @@ buildGoModule (finalAttrs: {
     runHook postInstallCheck
   '';
 
-  passthru.updateScript = nix-update-script { };
+  passthru = {
+    tests = { inherit (nixosTests) go-httpbin; };
+    updateScript = nix-update-script { };
+  };
 
   meta = {
     description = "Reasonably complete and well-tested golang port of httpbin, with zero dependencies outside the go stdlib";

